### ISSUE TITLE
reduceDimension error with UMAP option 

### ISSUE THREAD

**Issue by dwucsf** [cole-trapnell-lab/monocle-release#288]
I'm running monocle 3 (Version 2.99.3) and can't run the reduceDimension command using the UMAP option. 

cds <- reduceDimension(cds, reduction_method = 'UMAP')

The error I'm getting is:

Error in py_call_impl(callable, dots$args, dots$keywords) : 
  TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Unknown attribute 'shape' of type none

File "../../../../../../../../../anaconda/envs/r-reticulate/lib/python3.6/site-packages/umap/umap_.py", line 88:
def smooth_knn_dist(distances, k, n_iter=64, local_connectivity=1.0, bandwidth=1.0):
    <source elided>
    target = np.log2(k) * bandwidth
    rho = np.zeros(distances.shape[0])
    ^

[1] During: typing of get attribute at /Users/DanielWong/anaconda/envs/r-reticulate/lib/python3.6/site-packages/umap/umap_.py (88)

File "../../../../../../../../../anaconda/envs/r-reticulate/lib/python3.6/site-packages/umap/umap_.py", line 88:
def smooth_knn_dist(distances, k, n_iter=64, local_connectivity=1.0, bandwidth=1.0):
    <source elided>
    target = np.log2(k) * bandwidth
    rho = np.zeros(distances.shape[0])
    ^

This is not usually a problem with Numba itself but instead often caused by
the use of unsupported features or an issue in resolving types.

Anyone know what the problem is? When I run the function with the DDRTree option, it works fine, but I specifically want to use UMAP.

**Comment by FTD2018** [comment]
Hi, dwucsf !

This is a bug of numba 0.43 version which UMAP algorithm depended on. 
The solution to this problem is that you can use the numba 0.42 version instead of 0.43.

please see the link
[https://github.com/lmcinnes/umap/issues/209]

**Comment by lilei0202** [comment]
Hello,

I have the same problem and can not be solved by installing numba 0.42
Here is my R session:
```
sessionInfo()
R version 3.5.3 (2019-03-11)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Mojave 10.14.2

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] monocle_2.99.3        L1Graph_0.1.1         lpSolveAPI_5.5.2.0-17 DDRTree_0.1.5         irlba_2.3.3           igraph_1.2.4          ggplot2_3.1.0        
 [8] Biobase_2.42.0        DelayedArray_0.6.6    BiocParallel_1.14.2   IRanges_2.16.0        S4Vectors_0.20.1      BiocGenerics_0.28.0   matrixStats_0.54.0   
[15] Matrix_1.2-17        

loaded via a namespace (and not attached):
  [1] nlme_3.1-137             sf_0.7-3                 gmodels_2.18.1           doParallel_1.0.14        webshot_0.5.1            RColorBrewer_1.1-2      
  [7] httr_1.4.0               docopt_0.6.1             tools_3.5.3              R6_2.4.0                 spData_0.3.0             DBI_1.0.0               
 [13] lazyeval_0.2.2           colorspace_1.4-1         manipulateWidget_0.10.0  sp_1.3-1                 withr_2.1.2              tidyselect_0.2.5        
 [19] gridExtra_2.3            compiler_3.5.3           glmnet_2.0-16            expm_0.999-4             plotly_4.8.0             slam_0.1-45             
 [25] scales_1.0.0             classInt_0.3-1           pbapply_1.4-0            stringr_1.4.0            digest_0.6.18            sparsesvd_0.1-4         
 [31] pkgconfig_2.0.2          htmltools_0.3.6          limma_3.38.3             htmlwidgets_1.3          rlang_0.3.2              rstudioapi_0.10         
 [37] VGAM_1.1-1               DelayedMatrixStats_1.2.0 FNN_1.1.3                shiny_1.2.0              jsonlite_1.6             gtools_3.8.1            
 [43] crosstalk_1.0.0          spdep_1.0-2              dplyr_0.8.0.1            magrittr_1.5             Rcpp_1.0.1               munsell_0.5.0           
 [49] reticulate_1.11.1-9000   viridis_0.5.1            stringi_1.4.3            yaml_2.2.0               MASS_7.3-51.1            Rtsne_0.15              
 [55] plyr_1.8.4               grid_3.5.3               gdata_2.18.0             listenv_0.7.0            promises_1.0.1           ggrepel_0.8.0           
 [61] crayon_1.3.4             deldir_0.1-16            miniUI_0.1.1.1           lattice_0.20-38          splines_3.5.3            knitr_1.22              
 [67] pillar_1.3.1             boot_1.3-20              reshape2_1.4.3           codetools_0.2-16         LearnBayes_2.15.1        glue_1.3.1              
 [73] data.table_1.12.0        httpuv_1.5.0             foreach_1.4.4            gtable_0.2.0             RANN_2.6.1               purrr_0.3.2             
 [79] tidyr_0.8.3              future_1.12.0            assertthat_0.2.1         xfun_0.5                 mime_0.6                 xtable_1.8-3            
 [85] e1071_1.7-1              coda_0.19-2              later_0.8.0              class_7.3-15             HSMMSingleCell_1.2.0     qlcMatrix_0.9.7         
 [91] viridisLite_0.3.0        tibble_2.1.1             pheatmap_1.0.12          iterators_1.0.10         pbmcapply_1.3.1          units_0.6-2             
 [97] cluster_2.0.7-1          fastICA_1.2-1            densityClust_0.3         rgl_0.100.19             globals_0.12.4   
```
Numba version is 0.42
```
$ python
Python 3.7.1 (default, Dec 14 2018, 13:28:58) 
[Clang 4.0.1 (tags/RELEASE_401/final)] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numba
>>> numba.__version__
'0.42.0'
```
Conda list
```
# Name                    Version                   Build  Channel
asn1crypto                0.24.0                   py37_0  
blas                      1.0                         mkl  
ca-certificates           2019.1.23                     0  
certifi                   2019.3.9                 py37_0  
cffi                      1.11.5           py37h6174b99_1  
chardet                   3.0.4                    py37_1  
conda                     4.6.8                    py37_0  
conda-env                 2.6.0                         1  
cryptography              2.4.2            py37ha12b0ac_0  
idna                      2.8                      py37_0  
intel-openmp              2019.3                      199  
libcxx                    4.0.1                hcfea43d_1  
libcxxabi                 4.0.1                hcfea43d_1  
libedit                   3.1.20170329         hb402a30_2  
libffi                    3.2.1                h475c297_4  
libgfortran               3.0.1                h93005f0_2  
llvmlite                  0.27.0           py37h8c7ce04_0  
mkl                       2019.3                      199  
mkl_fft                   1.0.10           py37h5e564d8_0  
mkl_random                1.0.2            py37h27c97d8_0  
ncurses                   6.1                  h0a44026_1  
numba                     0.42.0           py37h6440ff4_0  
numpy                     1.16.2           py37hacdab7b_0  
numpy-base                1.16.2           py37h6575580_0  
openssl                   1.1.1b               h1de35cc_1  
pip                       18.1                     py37_0  
pycosat                   0.6.3            py37h1de35cc_0  
pycparser                 2.19                     py37_0  
pyopenssl                 18.0.0                   py37_0  
pysocks                   1.6.8                    py37_0  
python                    3.7.1                haf84260_7  
python.app                2                        py37_9  
readline                  7.0                  h1de35cc_5  
requests                  2.21.0                   py37_0  
ruamel_yaml               0.15.46          py37h1de35cc_0  
setuptools                40.6.3                   py37_0  
six                       1.12.0                   py37_0  
sqlite                    3.26.0               ha441bb4_0  
tk                        8.6.8                ha441bb4_0  
urllib3                   1.24.1                   py37_0  
wheel                     0.32.3                   py37_0  
xz                        5.2.4                h1de35cc_4  
yaml                      0.1.7                hc338f04_2  
zlib                      1.2.11               h1de35cc_3  
```

Could you help to figure out if there is anything wrong with my environment? 
Thanks!

**Comment by FTD2018** [comment]
Hi,
  Have you solved the problem yet? 

  Now the numba 0.43.1 is available now, you can re-install the umap again.
Xiaopeng Xu



------------------ 原始邮件 ------------------
发件人: "Lei Li"<notifications@github.com>;
发送时间: 2019年3月23日(星期六) 中午1:12
收件人: "cole-trapnell-lab/monocle-release"<monocle-release@noreply.github.com>;
抄送: "小徐"<912530451@qq.com>; "Comment"<comment@noreply.github.com>; 
主题: Re: [cole-trapnell-lab/monocle-release] reduceDimension error withUMAP option  (#288)




Hello,
 
I have the same problem and can not be solved by installing numba 0.42
 Here is my R session:
 sessionInfo() R version 3.5.3 (2019-03-11) Platform: x86_64-apple-darwin15.6.0 (64-bit) Running under: macOS Mojave 10.14.2 Matrix products: default BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib locale: [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8 attached base packages: [1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods   base      other attached packages:  [1] monocle_2.99.3        L1Graph_0.1.1         lpSolveAPI_5.5.2.0-17 DDRTree_0.1.5         irlba_2.3.3           igraph_1.2.4          ggplot2_3.1.0          [8] Biobase_2.42.0        DelayedArray_0.6.6    BiocParallel_1.14.2   IRanges_2.16.0        S4Vectors_0.20.1      BiocGenerics_0.28.0   matrixStats_0.54.0    [15] Matrix_1.2-17         loaded via a namespace (and not attached):   [1] nlme_3.1-137             sf_0.7-3                 gmodels_2.18.1           doParallel_1.0.14        webshot_0.5.1            RColorBrewer_1.1-2         [7] httr_1.4.0               docopt_0.6.1             tools_3.5.3              R6_2.4.0                 spData_0.3.0             DBI_1.0.0                 [13] lazyeval_0.2.2           colorspace_1.4-1         manipulateWidget_0.10.0  sp_1.3-1                 withr_2.1.2              tidyselect_0.2.5          [19] gridExtra_2.3            compiler_3.5.3           glmnet_2.0-16            expm_0.999-4             plotly_4.8.0             slam_0.1-45               [25] scales_1.0.0             classInt_0.3-1           pbapply_1.4-0            stringr_1.4.0            digest_0.6.18            sparsesvd_0.1-4           [31] pkgconfig_2.0.2          htmltools_0.3.6          limma_3.38.3             htmlwidgets_1.3          rlang_0.3.2              rstudioapi_0.10           [37] VGAM_1.1-1               DelayedMatrixStats_1.2.0 FNN_1.1.3                shiny_1.2.0              jsonlite_1.6             gtools_3.8.1              [43] crosstalk_1.0.0          spdep_1.0-2              dplyr_0.8.0.1            magrittr_1.5             Rcpp_1.0.1               munsell_0.5.0             [49] reticulate_1.11.1-9000   viridis_0.5.1            stringi_1.4.3            yaml_2.2.0               MASS_7.3-51.1            Rtsne_0.15                [55] plyr_1.8.4               grid_3.5.3               gdata_2.18.0             listenv_0.7.0            promises_1.0.1           ggrepel_0.8.0             [61] crayon_1.3.4             deldir_0.1-16            miniUI_0.1.1.1           lattice_0.20-38          splines_3.5.3            knitr_1.22                [67] pillar_1.3.1             boot_1.3-20              reshape2_1.4.3           codetools_0.2-16         LearnBayes_2.15.1        glue_1.3.1                [73] data.table_1.12.0        httpuv_1.5.0             foreach_1.4.4            gtable_0.2.0             RANN_2.6.1               purrr_0.3.2               [79] tidyr_0.8.3              future_1.12.0            assertthat_0.2.1         xfun_0.5                 mime_0.6                 xtable_1.8-3              [85] e1071_1.7-1              coda_0.19-2              later_0.8.0              class_7.3-15             HSMMSingleCell_1.2.0     qlcMatrix_0.9.7           [91] viridisLite_0.3.0        tibble_2.1.1             pheatmap_1.0.12          iterators_1.0.10         pbmcapply_1.3.1          units_0.6-2               [97] cluster_2.0.7-1          fastICA_1.2-1            densityClust_0.3         rgl_0.100.19             globals_0.12.4     
Numba version is 0.42
 $ python Python 3.7.1 (default, Dec 14 2018, 13:28:58)  [Clang 4.0.1 (tags/RELEASE_401/final)] :: Anaconda, Inc. on darwin Type "help", "copyright", "credits" or "license" for more information. >>> import numba >>> numba.__version__ '0.42.0'  
Conda list
 # Name                    Version                   Build  Channel asn1crypto                0.24.0                   py37_0   blas                      1.0                         mkl   ca-certificates           2019.1.23                     0   certifi                   2019.3.9                 py37_0   cffi                      1.11.5           py37h6174b99_1   chardet                   3.0.4                    py37_1   conda                     4.6.8                    py37_0   conda-env                 2.6.0                         1   cryptography              2.4.2            py37ha12b0ac_0   idna                      2.8                      py37_0   intel-openmp              2019.3                      199   libcxx                    4.0.1                hcfea43d_1   libcxxabi                 4.0.1                hcfea43d_1   libedit                   3.1.20170329         hb402a30_2   libffi                    3.2.1                h475c297_4   libgfortran               3.0.1                h93005f0_2   llvmlite                  0.27.0           py37h8c7ce04_0   mkl                       2019.3                      199   mkl_fft                   1.0.10           py37h5e564d8_0   mkl_random                1.0.2            py37h27c97d8_0   ncurses                   6.1                  h0a44026_1   numba                     0.42.0           py37h6440ff4_0   numpy                     1.16.2           py37hacdab7b_0   numpy-base                1.16.2           py37h6575580_0   openssl                   1.1.1b               h1de35cc_1   pip                       18.1                     py37_0   pycosat                   0.6.3            py37h1de35cc_0   pycparser                 2.19                     py37_0   pyopenssl                 18.0.0                   py37_0   pysocks                   1.6.8                    py37_0   python                    3.7.1                haf84260_7   python.app                2                        py37_9   readline                  7.0                  h1de35cc_5   requests                  2.21.0                   py37_0   ruamel_yaml               0.15.46          py37h1de35cc_0   setuptools                40.6.3                   py37_0   six                       1.12.0                   py37_0   sqlite                    3.26.0               ha441bb4_0   tk                        8.6.8                ha441bb4_0   urllib3                   1.24.1                   py37_0   wheel                     0.32.3                   py37_0   xz                        5.2.4                h1de35cc_4   yaml                      0.1.7                hc338f04_2   zlib                      1.2.11               h1de35cc_3    
Could you help to figure out if there is anything wrong with my environment?
 Thanks!

**Comment by dwucsf** [comment]
So sorry for the late reply! Been sorting out other lab work. Thank you so much fro following up @FTD2018!! Yes, I just re-installed UMAP and the reduceDimension function worked! Thanks again!
